cmd/kops-controller/controllers/gceipam.go

```diff
@@ -0,0 +1,150 @@
+	klog.Info("starting gce ipam controller")
+	r := &GCEIPAMReconciler{
+		client: mgr.GetClient(),
+		log:    ctrl.Log.WithName("controllers").WithName("gce-ipam"),
+	}
+
+	coreClient, err := corev1client.NewForConfig(mgr.GetConfig())
+	if err != nil {
+		return nil, fmt.Errorf("error building corev1 client: %w", err)
+	}
+	r.coreV1Client = coreClient
+
+	gceClient, err := compute.NewService(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("error building compute API client: %w", err)
+	}
+	r.gceClient = gceClient
+
+	return r, nil
+}
+
+// GCEIPAMReconciler observes Node objects, TODO
+type GCEIPAMReconciler struct {
+	// client is the controller-runtime client
+	client client.Client
+
+	// log is a logr
+	log logr.Logger
+
+	// coreV1Client is a client-go client for patching nodes
+	coreV1Client *corev1client.CoreV1Client
+
+	// gceClient is a client for GCE
+	gceClient *compute.Service
+}
+
+// +kubebuilder:rbac:groups=,resources=nodes,verbs=get;list;watch;patch
+// Reconcile is the main reconciler function that observes node changes.
+func (r *GCEIPAMReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = r.log.WithValues("node", req.NamespacedName)
+
+	node := &corev1.Node{}
+	if err := r.client.Get(ctx, req.NamespacedName, node); err != nil {
+		klog.Warningf("unable to fetch node %s: %w", node.Name, err)
+		if apierrors.IsNotFound(err) {
+			// we'll ignore not-found errors, since they can't be fixed by an immediate
+			// requeue (we'll need to wait for a new notification), and we can get them
+			// on deleted requests.
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	if len(node.Spec.PodCIDRs) == 0 {
+		// CCM Node Controller has not done its thing yet
+		if node.Spec.ProviderID == "" {
+			klog.Infof("node %q has empty provider ID", node.Name)
+			return ctrl.Result{}, nil
+		}
+
+		// e.g. providerID: gce://example-project-id/us-west2-a/instance-id
+		providerURL, err := url.Parse(node.Spec.ProviderID)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("error parsing providerID %q: %w", node.Spec.ProviderID, err)
+		}
+		tokens := strings.Split(strings.Trim(providerURL.Path, "/"), "/")
+		if len(tokens) != 2 {
+			return ctrl.Result{}, fmt.Errorf("unexpected format for providerID %q", node.Spec.ProviderID)
+		}
+		project := providerURL.Host
+		zone := tokens[0]
+		instanceID := tokens[1]
+		if project == "" || zone == "" || instanceID == "" {
+			return ctrl.Result{}, fmt.Errorf("unexpected format for providerID %q", node.Spec.ProviderID)
+		}
+
+		instance, err := r.gceClient.Instances.Get(project, zone, instanceID).Context(ctx).Do()
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("error getting instance %s/%s/%s: %w", project, zone, instanceID, err)
+		}
+
+		if len(instance.NetworkInterfaces) != 1 {
+			return ctrl.Result{}, fmt.Errorf("unexpected number of network interfaces for instance %q: %v", instanceID, len(instance.NetworkInterfaces))
```

Comment:
Would it make sense to iterate over the multiple NetworkInterfaces in case only one of them has an ExternalIpv6?

Reply:
Done!  I was worried that it wasn't clear what we should do here, but as long as we end up with a single IPv6 range, that's what really matters.